### PR TITLE
Add Edition view

### PR DESCRIPTION
**Pull request by boczeratul**
Edition information is very useful in wallets and marketplaces. As more and more projects on Flow mint multiple items with the same name, description and media file, it is beneficial to put this into standard.

For projects that have items with editions, they should implement this `Edition` view.

Screen Shot 2022-04-11 at 9 02 06 PM
Screen Shot 2022-04-11 at 9 03 57 PM

### CODE REVIEW

File: contracts/MetadataViews.cdc [hunk 1]
```diff
@@ -129,4 +129,26 @@ pub contract MetadataViews {
...
+
+        // The edition number of the object.
+        //
+        pub let editionNumber: UInt64
+
+        // The max edition number of this type of objects.
+        // 
+        // This field should only be provided for limited-editioned objects.
+        // 
+        pub let maxEdition: UInt64?
```
Comment (joshuahannan) [inline contracts/MetadataViews.cdc:147]: Can you add a little bit more detail in the comments about what edition number and maxEdition mean? I assume edition number is like a "24 of 100" type thing and maxEdition is 100 in that case, but that should be crystal clear from the comments and docs here

**Review by joshuahannan — COMMENTED** [review]
Just have the one comment so far! It looks good!

File: contracts/MetadataViews.cdc (hunk 1 above)
Comment (boczeratul) [reply contracts/MetadataViews.cdc]: updated.

**Review by psiemens — APPROVED** [review]
Thanks for adding tests and updating the examples!

**Review by satyamakgec — APPROVED** [review]
LGTM

**Comment by albeethekid** [comment]
No comments, other than this look great!

**Comment by joshuahannan** [comment]
@boczeratul We just merged another PR to master. Can you pull the latest master, `make test`, then push the resolved changes again?

**Comment by poiati** [comment]
Please read the issue: https://github.com/onflow/flow-nft/issues/58. There is an extensive discussion about editions and we need more thoughts on it. It is currently blocking this PR. Thank you!

File: contracts/MetadataViews.cdc [hunk 2]
```diff
@@ -238,4 +238,30 @@ pub contract MetadataViews {
...
+    // with the same name, description and media file. This information is 
+    // useful for wallets and marketplaces.
+    // 
+    pub struct Edition {
+
+        // The edition number of the object.
+        //
+        // For an "24 of 100 (#24/100)" item, the editionNumber is 24. 
+        //
+        pub let editionNumber: UInt64
```
Comment (austinkline) [inline contracts/MetadataViews.cdc:252]: We already know that the view is about editions, what's the reasoning for repeating that again with each variable name?

**Review by albeethekid — APPROVED** [review]

**Comment by bjartek** [comment]
The discussion in #58 has concluded so I think this can be merged?

**Comment by joshuahannan** [comment]
no, I'm working on a more sophisticated version of it that I'll have ready relatively soon

**Comment by bjartek** [comment]
I might have misunderstood something, but why do we need sophistication here? When we have Serial what do we need then except Edition(number:total)?

**Comment by joshuahannan** [comment]
A project might have multiple editions that it is a part of, so the view should probably support that. Something like this:
```
pub struct Edition {

        /// An arbitrary-sized list for any number of editions
        /// that the NFT might be a part of
        pub let infoList: [EditionInfo]

        init(_ infoList: [EditionInfo]) {
            self.infoList = infoList
        }
    }

    /// Edition information for a single edition
    pub struct EditionInfo {

        /// The name of the edition
        /// For example, this could be Set, Play, Series,
        /// or any other way a project could classify its editions
        pub let name: String?

        /// The edition number of the object.
        ///
        /// For an "24 of 100 (#24/100)" item, the number is 24. 
        ///
        pub let number: UInt64

        /// The max edition number of this type of objects.
        /// 
        /// This field should only be provided for limited-editioned objects.
        /// For an "24 of 100 (#24/100)" item, max is 100.
        /// For an item with unlimited edition, max should be set to nil.
        /// 
        pub let max: UInt64?

        init(name: String?, number: UInt64, max: UInt64?) {
            self.name = name
            self.number = number
            self.max = max
        }
    }
```

**Comment by joshuahannan** [comment]
@bluesign I think that information could be included in the name, right?

**Comment by bjartek** [comment]
Is it neccesary to pack a list if editions inside another type? A view of [Edition] is totally valid?

**Comment by joshuahannan** [comment]
Just following the same pattern that we used for royalties. When we say that the function can resolve a specific type, I like for the resolver to return an object of the actual type, not an array of objects of that type

**Comment by bjartek** [comment]
If we want to follow the example of Royalties can we then name the view Editions and the thing inside Edition?

**Comment by joshuahannan** [comment]
Good suggestions! New PR is in https://github.com/onflow/flow-nft/pull/88

**Comment by bluesign** [comment]
> @bluesign I think that information could be included in the name, right?

I am not sure tbh, Name looks like 'Set' here, it is more like a value ("Young Stars")

**Pull request closed without merging**